sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs

```diff
@@ -75,231 +65,231 @@ public BaseRequestBuilder<T> SetCompletion(OnErrorHandler errorHandler)
+            sw.Start();
 
-            // build request
-            request = (HttpWebRequest)HttpWebRequest.Create(url);
-            if (getFiles().Count == 0)
-            {
-                request.Timeout = client.Configuration.Timeout;
-            }
-            else
-            {
-                request.Timeout = Timeout.Infinite;
-            }
-            request.Method = "POST";
+            if (this.client == null) { this.Build(client); }
+            if (this.client == null) { OnComplete(null, new ClientException("No client instance defined")); }
 
-            request.Headers = getHeaders();
-            request.AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate;
-            request.Accept = "application/xml";
-            request.ContentType = getContentType();
+            var url = client.Configuration.ServiceUrl + "/api_v3" + getPath();
+
+            var requestId = (Interlocked.Increment(ref Client.REQUEST_COUNTER)).ToString("X5");
+            // Generate a unique task id to group logs
 
-            // Add proxy information if required
-            CreateProxy(request);
+            this.Log(string.Format("url : [{0}]", url), requestId);
 
-            request.BeginGetRequestStream(new AsyncCallback(RequestCallback), this);
+            var files = getFiles();
+            var request = BuildRequest(url, files, client.Configuration.Timeout);
+            await WriteRequestBodyAsync(files, request, requestId);
+            var responseObject = await GetResponseAsync(request, requestId);
+            this.Log(string.Format("execution time for ([{0}]: [{1}]", getPath(), sw.Elapsed), requestId);
+
+            return responseObject;
         }
 
-        public T ExecuteAndWaitForResponse(Client client = null)
+        public void Execute(Client client = null)
         {
-            var taskWrapper = Task.Run(() =>
+            var task = this.ExecuteAsync(client).ContinueWith(t =>
             {
-                var taskResult = new TaskCompletionSource<T>();
-                this.SetCompletion((result, err) =>
-                    {
-                        if (err != null)
-                        {
-                            taskResult.TrySetException(err);
-                        }
-                        else
-                        {
-                            taskResult.TrySetResult(result);
-                        }
-                    })
-                    .Execute(client);
-                return taskResult.Task;
-            });
-
-            taskWrapper.Wait();
-            return taskWrapper.Result;
+                if (t.Status == TaskStatus.Faulted) { OnComplete(null, t.Exception); return; }
+
+                var result = t.Result;
+                OnComplete(result, null);
+            }).ConfigureAwait(false);
         }
 
-        private void RequestCallback(IAsyncResult result)
+        public T ExecuteAndWaitForResponse(Client client = null)
```

Comment:
You can rename to ExecuteSync.

Reply:
while i completely agree with you, I can't this was already there and will require all to change names .. backward comparability